File: examples/pki/simple_example.cpp
#include <argu/argu.hpp>
#include <authbox.hpp>
#include <echo/echo.hpp>
#include <scan/scan.hpp>

int main(int argc, char *argv[]) {
    authbox::log_startup();

    // Variables to bind arguments to
    std::string name;
    int count = 1;
    std::vector<std::string> files;
    std::string did_uri = "did:web:example.com";

    // Build the command
    auto cmd = argu::Command("argue_basic")
                   .version("1.0.0")
                   .about("A basic example demonstrating the Argue argument parsing library")
                   .arg(argu::Arg("name").positional().help("Your name").required().value_of(name).value_name("NAME"))
                   .arg(argu::Arg("count")
                            .short_name('c')
                            .long_name("count")
                            .help("Number of times to greet")
                            .value_of(count)
                            .default_value("1")
                            .value_name("NUM"))
                   .arg(argu::Arg("files")
                            .short_name('f')
                            .long_name("file")
                            .help("Input files to process")
                            .value_of(files)
                            .takes_multiple()
                            .value_name("FILE"))
                   .arg(argu::Arg("did")
                            .long_name("did")
                            .help("Optional DID URI to validate")
                            .value_of(did_uri)
                            .default_value("did:web:example.com")
                            .value_name("DID"));

    auto result = cmd.parse(argc, argv);
    if (!result) {
        return result.exit();
    }

    for (int i = 0; i < count; ++i) {
        echo("Hello, ", name, "!");
    }

    if (!files.empty()) {
        std::cout << "\nWould process files: ";
        for (size_t i = 0; i < files.size(); ++i) {
            if (i > 0)
                echo(", ");
            echo(files[i]);
        }
        echo("\n");
    }

    auto did_res = authbox::did::parse(did_uri);
    if (!did_res.is_ok()) {
        echo::warn("Invalid DID: ", did_uri);
    } else {
        echo::info("Parsed DID method=", did_res.value().method.c_str());
    }

    return 0;
}

File: examples/rust/pki/simple_example.rs
use authbox::{did, log_startup};

#[derive(Debug)]
struct Args {
    name: String,
    count: usize,
    files: Vec<String>,
    did_uri: String,
}

fn main() -> Result<(), Box<dyn std::error::Error>> {
    log_startup();

    let args = parse_args()?;
    for _ in 0..args.count {
        println!("Hello, {}!", args.name);
    }

    if !args.files.is_empty() {
        println!("\nWould process files: {}", args.files.join(", "));
    }

    match did::parse(&args.did_uri) {
        Ok(parsed) => println!("Parsed DID method={}", parsed.method),
        Err(_) => eprintln!("Invalid DID: {}", args.did_uri),
    }

    Ok(())
}

fn parse_args() -> Result<Args, Box<dyn std::error::Error>> {
    let mut name = None;
    let mut count = 1usize;
    let mut files = Vec::new();
    let mut did_uri = "did:web:example.com".to_string();

    let mut args = std::env::args().skip(1);
    while let Some(arg) = args.next() {
        match arg.as_str() {
            "-c" | "--count" => {
                let value = args
                    .next()
                    .ok_or_else(|| format!("{arg} requires a NUM value"))?;
                count = value
                    .parse::<usize>()
                    .map_err(|_| format!("{arg} requires a positive integer"))?;
            }
            "-f" | "--file" => {
                let value = args
                    .next()
                    .ok_or_else(|| format!("{arg} requires a FILE value"))?;
                files.push(value);
            }
            "--did" => {
                did_uri = args
                    .next()
                    .ok_or_else(|| "--did requires a DID value".to_string())?;
            }
            "-h" | "--help" => {
                print_usage();
                std::process::exit(0);
            }
            value if value.starts_with('-') => {
                return Err(format!("Unknown flag: {value}").into());
            }
            value => {
                if name.is_some() {
                    return Err(format!("Unexpected extra positional argument: {value}").into());
                }
                name = Some(value.to_string());
            }
        }
    }

    Ok(Args {
        name: name.unwrap_or_else(|| "friend".to_string()),
        count,
        files,
        did_uri,
    })
}

fn print_usage() {
    println!("Usage: simple_example [NAME] [-c NUM] [-f FILE] [--did DID]");
}
